_posts/minh_ha/2018-06-22-Tensorflow-fizz-buzz.md

```diff
@@ -0,0 +1,590 @@
+
+```javascript
+for (var i = 0; i < 100; i++) {
+  if (i % 15 === 0) console.log("FizzBuzz");
+  else if (i % 3 === 0) console.log("Fizz");
+  else if (i % 5 === 0) console.log("Buzz");
+  else console.log(i);
+}
+```
+
+... and it almost reads as if itself is the description of the game.
+But it's got looping and conditions, all the good stuff that traditional software
+boils down to. And in most programming languages, the code will continue to look
+mostly the same.
+
+### An entry-level interview question? Must be a piece of cake to solve for an almighty AI to solve!
+
+# Let's jump right in!
+
+The first thing I needed to learn was how ML works differently from traditional
+programming. For this, I found the following diagram to be a great summary.
+
+<image src="{{ site.urlimg }}/minh_ha/ml_difference.png" />
+
+For my Fizz buzz program, this means the output of my code is not the Fizz buzz
+sequence, but a program that can by itself, figure out how to solve Fizz buzz
+(already, this seems to ring very well with the notion of ML being "machines writing
+their own code"). This also mean, we needed to establish a ground rule: I am not
+allowed to tell the machine what to do explicitly, but rather let it learn by itself
+how to solve the problem! (This turns out to be more challenging for my monkey
+brain to follow than expected).
+
+The process for doing something like this can be summarized as:
+
+  1. Create training data
+  2. Set up the model
+  3. Train the model with the training data from `step 1`
+  4. Use the model to run inference (aka, watch it magically demonstrate its news
+   math skills an an epic Fizz buzz series)
+
+### So I followed the steps...
+
+1) **Create training data:**
+   What we will need here is a list of inputs and the corresponding output we want
+   our program to give for each input. In other words, a set of numbers, and the
+   corresponding Fizz buzz output we expect the program to give. This turned out to
+   be not exactly the most readily available dataset in the world, so I am simply going
+   to generate this with the code we have earlier, and pretend I just found it after
+   an extensive research in a lab... something like this should work:
+
+   _(if you see the problem with this, please feel free to skip this section...
+   or read on if you enjoy seeing me fail)_
+
+   ```
+   [
+     { "x": 0, "y": "fizzBuzz" },
+     { "x": 1, "y": 1 },
+     { "x": 2, "y": 2 },
+     { "x": 3, "y": "fizz" },
+     ...
+     { "x": 120, "y": "fizzBuzz" }
+   ]
+   ```
+
+2) **Set up the model:**
+   I went ahead and set up an [artificial neural network](https://en.wikipedia.org/wiki/Artificial_neural_network),
+   the spooky "simulation of a human brain" that has been in the news. With Tensorflow,
+   the code may look something like so:
+
+   ```javascript
+     const fizzBuzzModel = tf.sequential();
+     fizzBuzzModel.add(tf.layers.dense({ units: 10, inputShape: 1 }));
```

Comment:
At that point, I wonder what "relu" and "softmax" are or mean. I see those are listed in the above wikipedia page, but maybe a very brief sentence, not necessarily on what both are, but on what is "activation".

Reply:
Thanks for the feedback @Fgerthoffert :) I've added some clarification and links to resources